### ISSUE TITLE
Create Example

### ISSUE THREAD

**Issue by arturh85** [Ekleog/rlua-async#1]
I have been trying to use this crate but after a few days bashing my head in I have created a repository with a minimal example here: https://github.com/arturh85/rlua-async-example

The Problem is executing my lua code with chunk.exec_async:

```
lua.context(|lua_context| async move {
        let chunk = lua_context
            .load(&lua_code);
        chunk.exec_async(lua_context).await.unwrap();
    })
        .await;
```

But I'm getting this error message:

```
error: lifetime may not live long enough
  --> src\main.rs:28:31
   |
28 |       lua.context(|lua_context| async move {
   |  __________________------------_^
   | |                  |          |
   | |                  |          return type of closure is impl Future
   | |                  has type `LuaContext<'1>`
29 | |         let chunk = lua_context
30 | |             .load(&lua_code);
31 | |         chunk.exec_async(lua_context).await.unwrap();
32 | |     })
   | |_____^ returning this value requires that `'1` must outlive `'2`
```

Could you provide a usage example for this crate please?

**Comment by Ekleog** [comment]
I have actually never tried using rlua-async with actix yet, so this should probably be taken with a grain of salt, but would it be possible for you to refactor the code so you have the `lua.context(|lua_ctx| { /* the code here */ })` actually be outside the executor (ie. async_rt::main)?

The reason is, `lua.context()` isn't async-ready (I haven't found a way to make it async yet, though I must say it's not the thing I've been thinking most about), so you'd have to have something like `lua.context(|lua_ctx| actix::execute_main(async || { /* the code here */ }))` where `actix::execute_main` is a hypothetical equivalent to `#[actix_rt::main]`.

If you have an idea for how to async-ify `lua.context`, I'd love to hear it!

As for the original issue of creating an example, for now all there is are the tests around the bottom of `src/lib.rs`, so I'll keep this issue open even once your specific question is solved to not forget actually doing it — actually… if that's ok with you, maybe I'll even cheat and reuse your code, if you're OK with it coming in this repository :)

**Comment by arturh85** [comment]
I have created a version of my minimal example which uses block_on instead like in your unit tests: https://github.com/arturh85/rlua-async-example/blob/block_on/src/main.rs

```
    lua.context(|lua_context| {
        actix::System::new("Default")
            .block_on(lua_context.load(LUA_CODE).exec_async(lua_context))
            .unwrap();
    });
```

But this also gives an error which points to your observation that lua.context is just not async-ready:

```
error[E0495]: cannot infer an appropriate lifetime for lifetime parameter `'lua` due to conflicting requirements
  --> src\main.rs:22:35
   |
22 |             .block_on(lua_context.load(LUA_CODE).exec_async(lua_context))
   |                                   ^^^^
   |
note: first, the lifetime cannot outlive the anonymous lifetime #2 defined on the body at 20:17...
  --> src\main.rs:20:17
   |
20 |       lua.context(|lua_context| {
   |  _________________^
21 | |         actix::System::new("Default")
22 | |             .block_on(lua_context.load(LUA_CODE).exec_async(lua_context))
23 | |             .unwrap();
24 | |     });
   | |_____^
note: ...so that the types are compatible
  --> src\main.rs:22:35
   |
22 |             .block_on(lua_context.load(LUA_CODE).exec_async(lua_context))
   |                                   ^^^^
   = note: expected `LuaContext<'_>`
              found `LuaContext<'_>`
   = note: but, the lifetime must be valid for the static lifetime...
note: ...so that the type `Pin<Box<dyn Future<Output = std::result::Result<(), LuaError>>>>` will meet its required lifetime bounds
  --> src\main.rs:22:14
   |
22 |             .block_on(lua_context.load(LUA_CODE).exec_async(lua_context))
   |              ^^^^^^^^
```

I'm not sure how to specify the lifetime of 'lua as it is a lifetime of the return value of load:

```
pub fn load<'a, S>(self, source: &'a S) -> Chunk<'lua, 'a>
```

Any ideas? I'm sorry but I'm quite new to rust.

**Comment by Ekleog** [comment]
No problem! This, to the best of my understanding, is a problem in actix, which I've reported as https://github.com/actix/actix/issues/429.

Hopefully this will help you making things work! Once that's fixed on the actix side, your code should probably look something like this:
```
    lua.context(|lua_context| {
        actix::System::new("Default")
            .block_on(async move {
                // foo
                lua_context.load(LUA_CODE).exec_async(lua_context)
                // bar
            })
            .unwrap();
    });
```

Now, one possible hack to work around this actix limitation may be to use something like this, to artificially expand the lifetime of `lua_context` to `'static` (ie. live for the whole program instead of just the duration of the actual variable):
```
    let lua = /* ... */;
    let lua = Box::leak(Box::new(lua)); // Make lua be 'static
    lua.context(|lua_context| { /* see above */ });
```
(where you *may* need to leak `lua_context` too, I'm not entirely sure about this, but I'd say probably not)

Does this sound reasonable enough to you?

**Comment by Ekleog** [comment]
Actually, upon some more thought, I don't think the hack would actually work, because `lua_context` would still have the branding `'lua` lifetime by design of rlua, despite the fact that the `lua` variable would indeed be `'static` (this being by design because rlua has to make sure that the lifetime is correct, though it's one of the parts of rlua I understand least).

So… I'd say, you'll probably have to wait for actix to remove that bound, unfortunately… or at least I can't think of another workaround to avoid running into this `'static` bound there is on `block_on`, sorry :/

Now, I see this function in actix https://docs.rs/actix/0.10.0/actix/prelude/struct.SystemRunner.html#method.run that looks like it doesn't have that 'static bound, so maybe using this function would allow you to do this, by making `lua` be in a `lazy_static` or similar? Though, with the issue you linked above, you'd probably have to put `lua_context` in a `scoped_tls` or similar, and I'm not really sure yet how that'd be possible as it's not `'static`

**Comment by arturh85** [comment]
Wow, thank you a lot, I have tried to Box::leak all the things and even tried to wrap unsafe blocks around it but the error message does not change :(

The SystemRunner run method does not have any parameters, how would I use it to run the future without block_on? 

I'll watch what happens in the actix issue. Maybe I need to switch to tokio, which would not be that easy considering that my main project uses actix web, but I just googled and it is at least possible.

**Comment by arturh85** [comment]
I got it working in the minimal example and in my project by switching to tokio :-)
Here is my working code:

```
use rlua::Lua;
use rlua_async::{ChunkExt, ContextExt};

fn main() -> rlua::Result<()> {
    let lua = Lua::new();
    lua.context(|ctx| {
        let globals = ctx.globals();
        let map_table = ctx.create_table()?;
        map_table.set(
            "asyncfunc",
            ctx.create_async_function(|_ctx, param: u32| async move {
                println!("async function called {}", param);
                Ok(())
            })?,
        )?;
        globals.set("my", map_table)?;
        let chunk = ctx.load("my.asyncfunc(42)");
        tokio::runtime::Runtime::new()
            .unwrap()
            .block_on(chunk.exec_async(ctx))
    })?;
    Ok(())
}
```

In my real project I was already inside of an async executor context so I had to start a new thread but everything works fine then.
Thanks again for your kind help! You may of course reuse the code as you wish for documentation purposes.

**Comment by Ekleog** [comment]
Cool, thank you for your working code! Do you want to submit a PR for it, so that your username and email enter the git history (and thus are present in github statistics, etc.), or should I just copy-paste the code as an example, thus losing authorship information?

**Comment by arturh85** [comment]
I created a pull request for you: https://github.com/Ekleog/rlua-async/pull/2
Hope this is ok?

**Comment by Ekleog** [comment]
It's great, thank you!

**Comment by MindPatch** [comment]
hey @arturh85 @Ekleog , do you guys have another alternative for the `block_on` function?
because I'm working on another project that runs 5 lua scripts at the same time and when I tried to do this with futures I noticed that the scripts are working on one synchronously because the `block_on` function is blocking the thread until the process end

https://github.com/rusty-sec/lotus/blob/master/src/lib.rs#L50

there's any solution for this ?

best regards

**Comment by arturh85** [comment]
you should be able to start each in its own thread and join them together to wait for completion.

**Comment by MindPatch** [comment]
@arturh85 I tried this before but I got this error 
```
`dyn futures::Future<Output = Result<(), LuaError>>` cannot be sent between threads safely

```

**Comment by arturh85** [comment]
i get bad flashbacks seeing this particular error again :D I solved it with more threads, (and more async runtimes?) until it worked... i wasnt very happy with the result: https://github.com/arturh85/factorio-bot/blob/master/crates/scripting_lua/src/lua_runner.rs starting at line 67, the trick was to start the thread only at the end

**Comment by Ekleog** [comment]
@knassar702 Just so I understand correctly, do you want to run multiple lua threads in parallel, or do you want to run lua threads asynchronously?

The goal of this crate is to make it possible to have async host functions provided to lua code, and execute lua code in a way that it will not block the executor thread while running these host functions.

If you have multiple lua functions that you wish to run concurrently, it is another problem that this crate is not designed to solve. In particular, `rlua::Context` is neither Send nor Sync, so you probably will not be able to do it without creating one rlua::Context per executor thread.

Basically, this crate works best with thread-local executors, and not that well with work-stealing executors, because of limitations in rlua, that are themselves because of limitations in lua itself.
